### PR TITLE
Feature/8330 signature check/8694 notifications

### PR DESCRIPTION
**Pull request by pascal-brause**
## Description
Handles local notifications for the validity stats `expiringSoon` and `expired` of vaccination and recovery certificates.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8694

## Screenshots
8694-notification

### CODE REVIEW

**Review by NataliaLemmerth — APPROVED** [review]
UA approved.

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift [hunk 1]
```diff
@@ -287,6 +287,14 @@ class AppDelegate: UIResponder, UIApplicationDelegate, CoronaWarnAppDelegate, Re
 		)
 		#endif
 	}()
+	
+	lazy var healthCertificateService: HealthCertificateService = HealthCertificateService(
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift:291]:
```suggestion
	private lazy var healthCertificateService: HealthCertificateService = HealthCertificateService(
```

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/NotificationManager.swift [hunk 2]
```diff
@@ -10,10 +10,20 @@ final class NotificationManager: NSObject, UNUserNotificationCenterDelegate {
...
-		// Checkout a event checkin.
-		if notification.request.identifier.contains(EventCheckoutService.notificationIdentifierPrefix) {
+		// Show badge on check-in tab when checkout.
+		if notification.request.identifier.contains(LocalNotificationIdentifier.checkout.rawValue) {
 			appDelegate?.eventCheckoutService.checkoutOverdueCheckins()
 		}
+		
+		// Show badge on certificates tab when certificate is expired.
+		if notification.request.identifier.contains(LocalNotificationIdentifier.certificateExpired.rawValue) {
+			appDelegate?.healthCertificateService.unseenTestCertificateCount.value += 1
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/NotificationManager.swift:20]: why not overwrite init and inject healthCertificateService?
Also you can restructure the class and add some MARKs as well.

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift [hunk 3]
```diff
@@ -621,6 +644,125 @@ class HealthCertificateService {
...
+		let expirationThresholdInDays = appConfiguration.currentAppConfig.value.dgcParameters.expirationThresholdInDays
+		let expiringSoonDate = Calendar.current.date(
+			byAdding: .day,
+			value: -Int(expirationThresholdInDays),
+			to: healthCertificate.expirationDate
+		)
+		
+		let expirationDate = healthCertificate.expirationDate
+		self.scheduleNotificationForExpiringSoon(id: id, date: expiringSoonDate)
+		self.scheduleNotificationForExpired(id: id, date: expirationDate)
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift:690]:
```suggestion
		scheduleNotificationForExpiringSoon(id: id, date: expiringSoonDate)
		scheduleNotificationForExpired(id: id, date: expirationDate)
```

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift [hunk 4]
```diff
@@ -493,9 +499,26 @@ class HealthCertificateService {
...
 				}
+				
+				self.updateNotifications()
 			}
 			.store(in: &subscriptions)
 	}
+	
+	/// This method should be called: At startup, at creation, at removal and at update validity states of HealthCertificates.
+	/// First, removes all local notifications and then re-adds all updates or new notifications to the notification center.
+	func updateNotifications() {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift:510]: I only see the call for the case of updating validity states.
Is there a task to call this func in the other cases? So we don't forget to wire this up?

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift (hunk 1 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/AppDelegate & Globals/AppDelegate.swift:291]: re-added `private` after now injecting the service to the notification manager

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/NotificationManager.swift (hunk 2 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/AppDelegate & Globals/NotificationManager.swift:20]: - Injected all services and removed the reference to the appDelegate.
- Removed obsolete function
- Added MARKs

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift (hunk 4 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift:510]: The update shall be triggered when the app config changes or the validity changes or at startup. In every of these cases, `updateValidityStatesAndNotifications` was already called so i added the `updateNotifications` just at the end of the `updateValidityStatesAndNotifications` function to cover these three scenarios.

**Review by KaiTeuber — APPROVED** [review]

**Review by ArturFriesen — APPROVED** [review]

**Pull request merged**